### PR TITLE
Implement draft-perez-krb-wg-gss-preauth-02.txt

### PR DESCRIPTION
**Pull request by lhoward**
Creating PR to track GSS-API pre-authentication support

Note this protocol differs from the draft: GSS pre-auth contains the raw GSS-API token (not encapsulated in PA-GSS) and the state is managed by FAST FX-COOKIE.

### CODE REVIEW

File: kadmin/init.c [hunk 1]
```diff
@@ -246,6 +252,15 @@ init(struct init_options *opt, int argc, char **argv)
 			    KRB5_KDB_DISALLOW_ALL_TIX, CRE_DUP_OK);
 	krb5_free_principal(context, princ);
 
+	/* Create `WELLKNONW/org.h5l.gss-preauth-state@WELLKNOWN:ORG.H5L' for preauth state */
```
Comment (lhoward) [inline kadmin/init.c:255]: This principal is used to encrypt acceptor state, so the acceptor is stateless. We could just use the FAST cookie principal, given they have separate key usage numbers?

**Comment by elric1** [comment]
It would be quite nice to add tracing on the new code that we add. For the KDC, that's level >= 7 on `kdc_log()`. In libgssapi, it's `_gss_mg_log()`. Looks like I used level 10, but I should have used level 7 based on the documentation in `krb5_openlog(3)`.

**Comment by lhoward** [comment]
Noted. A lot of the Apple code traced at level 0, but I've dialled it back to 4 or 5 and fixed a bunch of typos / spelling errors. It's pretty verbose though, could probably be refined. e.g. here's getting an anonymous ticket with FAST armour and SAnon pre-authentication inside NegoEx ;)

```bash
% ./kinit --fast-armor-cache=armor --gss-mech=spnego --anonymous
2021-08-10T14:06:26 No such directory /Users/lukeh/CVSRoot/heimdal/lib/base/.libs/../lib/plugin/krb5
2021-08-10T14:06:26 No such directory /usr/local/heimdal/lib/plugin/krb5
2021-08-11T00:06:26 gss_acquire_cred sanon-x25519 name: 0/0 no name
2021-08-11T00:06:26 gss_acquire_cred_from: sanon-x25519
2021-08-11T00:06:26 krb5_get_creds: requesting a ticket for krbtgt/PA.PADL.COM@PA.PADL.COM
2021-08-11T00:06:26 Adding PA mech: GSS
2021-08-11T00:06:26 krb5_get_init_creds: loop 1
2021-08-11T00:06:26 krb5_get_creds: requesting a ticket for krbtgt/PA.PADL.COM@PA.PADL.COM
2021-08-11T00:06:26 KDC sent 0 patypes
2021-08-11T00:06:26 pa-mech trying: GSS, searching for 655
2021-08-11T00:06:26 Stepping pa-mech: GSS
2021-08-11T00:06:26 gss_isc: spnego first flags 00000042, specific cred, no token
2021-08-11T00:06:26 gss_isc cred: sanon-x25519
2021-08-11T00:06:26 gss_isc: sanon-x25519 first flags 00000042, specific cred, no token
2021-08-11T00:06:26 gss_isc cred: sanon-x25519
2021-08-11T00:06:26 gss_isc: sanon-x25519 maj_stat: 1/0
2021-08-11T00:06:26 spnego: initiator proposed mech negoex { 1 3 6 1 4 1 311 2 2 30 }
2021-08-11T00:06:26 spnego: initiator proposed mech sanon-x25519 { 1 3 6 1 4 1 5322 26 1 110 }
2021-08-11T00:06:26 gss-qmd: return 0/0
2021-08-11T00:06:26 _gss_mg_error: captured unknown mech-code 2529638957 for mech 1 2 840 113554 1 2 2 (-1765328339) from underlying mech sanon-x25519
2021-08-11T00:06:26 _gss_mg_error: captured unknown mech-code 2529638957 for mech 1 2 840 113554 1 2 2 (-1765328339) from underlying mech sanon-x25519
2021-08-11T00:06:26 negoex: proposing authentication scheme 1 dee384ff-1086-4e86-be78-b94170bfd376
2021-08-11T00:06:26 negoex: sending (0)INITIATOR_NEGO conversation e6b9604e-8e77-dd2c-6397-c096959569be
2021-08-11T00:06:26 negoex: sending (1)AP_REQUEST conversation e6b9604e-8e77-dd2c-6397-c096959569be
2021-08-11T00:06:26 gss_isc: spnego maj_stat: 1/0
2021-08-11T00:06:26 PA type GSS returned -1980176627
2021-08-11T00:06:26 Continue needed for GSS
2021-08-11T00:06:26 pamech need more stepping
2021-08-11T00:06:26 Trying to find service kdc for realm PA.PADL.COM flags 0
2021-08-11T00:06:26 configuration file for realm PA.PADL.COM found
2021-08-11T00:06:26 submitting new requests to new host
2021-08-11T00:06:26 connecting to host: tcp ::1:accessbuilder (localhost) state=CONNECT tid: 00000001
2021-08-11T00:06:26 connecting to 3: tcp ::1:accessbuilder (localhost) state=CONNECT tid: 00000001
2021-08-11T00:06:26 Queuing host in future (in 3s), its the 2 address on the same name: tcp 127.0.0.1:accessbuilder (localhost) state=CONNECT tid: 00000002
2021-08-11T00:06:26 writing packet: tcp ::1:accessbuilder (localhost) state=CONNECTED tid: 00000001
2021-08-11T00:06:26 reading packet: tcp ::1:accessbuilder (localhost) state=WAITING_REPLY tid: 00000001
2021-08-11T00:06:26 host completed: tcp ::1:accessbuilder (localhost) state=WAITING_REPLY tid: 00000001
2021-08-11T00:06:26 unknown-function PA.PADL.COM done: 0 hosts: 1 packets: 1 wc: 0.015158 nr: 0.007314 kh: 0.000052 tid: 00000002
2021-08-11T00:06:26 krb5_get_init_creds: loop 2
2021-08-11T00:06:26 krb5_get_init_creds: processing input
2021-08-11T00:06:26 krb5_get_init_creds: got an KRB-ERROR from KDC
2021-08-11T00:06:26 krb5_get_init_creds: KRB-ERROR -1765328293/More pre-authentication data required
2021-08-11T00:06:26 krb5_get_creds: requesting a ticket for krbtgt/PA.PADL.COM@PA.PADL.COM
2021-08-11T00:06:26 KDC sent 4 patypes
2021-08-11T00:06:26 KDC sent PA-DATA type: 655 (GSS)
2021-08-11T00:06:26 KDC sent PA-DATA type: 19 (ETYPE_INFO2)
2021-08-11T00:06:26 KDC sent PA-DATA type: 133 (FX-COOKIE)
2021-08-11T00:06:26 KDC sent PA-DATA type: 137 (FX_ERROR)
2021-08-11T00:06:26 pa-mech trying: GSS, searching for 655
2021-08-11T00:06:26 Stepping pa-mech: GSS
2021-08-11T00:06:26 gss_isc: spnego not first flags 00000042, specific cred, token
2021-08-11T00:06:26 gss_isc cred: sanon-x25519
2021-08-11T00:06:26 spnego: initiator selected mechanism negoex { 1 3 6 1 4 1 311 2 2 30 }
2021-08-11T00:06:26 negoex: received (2)ACCEPTOR_NEGO conversation e6b9604e-8e77-dd2c-6397-c096959569be
2021-08-11T00:06:26 negoex: received (3)CHALLENGE conversation e6b9604e-8e77-dd2c-6397-c096959569be
2021-08-11T00:06:26 negoex: received (4)VERIFY conversation e6b9604e-8e77-dd2c-6397-c096959569be
2021-08-11T00:06:26 gss_isc: sanon-x25519 not first flags 00000042, specific cred, token
2021-08-11T00:06:26 gss_isc cred: sanon-x25519
2021-08-11T00:06:26 gss_isc: sanon-x25519 maj_stat: 0/0
2021-08-11T00:06:26 negoex: sending (5)VERIFY conversation e6b9604e-8e77-dd2c-6397-c096959569be
2021-08-11T00:06:26 _gss_mg_error: captured unknown mech-code 22 for mech 1 2 840 113554 1 2 2 (22) from underlying mech sanon-x25519
2021-08-11T00:06:26 _gss_mg_error: captured unknown mech-code 22 for mech 1 2 840 113554 1 2 2 (22) from underlying mech sanon-x25519
2021-08-11T00:06:26 spnego: mechListMIC omitted as preferred mechanism selected
2021-08-11T00:06:26 gss_isc: spnego maj_stat: 1/0
2021-08-11T00:06:26 PA type GSS returned -1980176627
2021-08-11T00:06:26 Continue needed for GSS
2021-08-11T00:06:26 pamech need more stepping
2021-08-11T00:06:26 Mirrored FX-COOKIE to KDC
2021-08-11T00:06:26 Trying to find service kdc for realm PA.PADL.COM flags 2
2021-08-11T00:06:26 configuration file for realm PA.PADL.COM found
2021-08-11T00:06:26 submitting new requests to new host
2021-08-11T00:06:26 connecting to host: tcp ::1:accessbuilder (localhost) state=CONNECT tid: 00010001
2021-08-11T00:06:26 connecting to 3: tcp ::1:accessbuilder (localhost) state=CONNECT tid: 00010001
2021-08-11T00:06:26 Queuing host in future (in 3s), its the 2 address on the same name: tcp 127.0.0.1:accessbuilder (localhost) state=CONNECT tid: 00010002
2021-08-11T00:06:26 writing packet: tcp ::1:accessbuilder (localhost) state=CONNECTED tid: 00010001
2021-08-11T00:06:26 reading packet: tcp ::1:accessbuilder (localhost) state=WAITING_REPLY tid: 00010001
2021-08-11T00:06:26 host completed: tcp ::1:accessbuilder (localhost) state=WAITING_REPLY tid: 00010001
2021-08-11T00:06:26 unknown-function PA.PADL.COM done: 0 hosts: 1 packets: 1 wc: 0.011941 nr: 0.007112 kh: 0.000010 tid: 00010002
2021-08-11T00:06:26 krb5_get_init_creds: loop 3
2021-08-11T00:06:26 krb5_get_init_creds: processing input
2021-08-11T00:06:26 KDC sent 1 patypes
2021-08-11T00:06:26 KDC sent PA-DATA type: 655 (GSS)
2021-08-11T00:06:26 pa-mech trying: GSS, searching for 655
2021-08-11T00:06:26 Stepping pa-mech: GSS
2021-08-11T00:06:26 gss_isc: spnego not first flags 00000042, specific cred, token
2021-08-11T00:06:26 gss_isc cred: sanon-x25519
2021-08-11T00:06:26 gss_isc: spnego maj_stat: 0/0
2021-08-11T00:06:26 pa_gss_step: WELLKNOWN/FEDERATED@PA.PADL.COM as WELLKNOWN/ANONYMOUS@WELLKNOWN:ANONYMOUS
2021-08-11T00:06:26 PA type GSS returned 0
2021-08-11T00:06:26 final pamech done step
2021-08-11T00:06:26 krb5_get_init_creds: FAST strengthen_key
2021-08-11T00:06:26 krb5_get_init_creds: extracting ticket
2021-08-11T00:06:26 processing enc-ap-rep
2021-08-11T00:06:26 enc-ap-rep: valid
2021-08-11T00:06:26 krb5_get_init_creds: wc: 0.031898
```

**Comment by lhoward** [comment]
I can add some GSS pre-auth specific tracing, though.

**Comment by elric1** [comment]
I think that detailed messages are fine as long as the levels are quite high. After all, you only turn it up to 10 when something is going wrong and you can't work out what it is. At this point, too many messages within reason is better than too few.

**Comment by lhoward** [comment]
> I think that detailed messages are fine as long as the levels are quite high. After all, you only turn it up to 10 when something is going wrong and you can't work out what it is. At this point, too many messages within reason is better than too few.

Good point, thanks for encouraging this, I've added a bunch of debugging to `kdc/gss_preauth.c`. I think, at least by the time we get around to merging #805, the client side logging is good.

File: kdc/gss_preauth.c [hunk 2]
```diff
@@ -0,0 +1,804 @@
...
+pa_gss_get_context_state(astgs_request_t r,
+                         gss_client_params *gcp)
+{
+    int idx = 0;
+    PA_DATA *fast_pa;
+
+    fast_pa = krb5_find_padata(r->fast.fast_state.val,
+                               r->fast.fast_state.len,
+                               KRB5_PADATA_GSS, &idx);
+    if (fast_pa) {
```
Comment (nicowilliams) [inline kdc/gss_preauth.c:77]: If `!fast_pa` we return zero, but that seems wrong.

File: lib/krb5/init_creds_pw.c [hunk 3]
```diff
@@ -1169,6 +1198,105 @@ pa_data_to_md_pkinit(krb5_context context,
...
+		       &ctx->as_req.req_body, &len, ret);
+    if (ret)
+	goto out;
+    heim_assert(req_body.length == len, "ASN.1 internal error");
+
+    ret = gssic->step(context, gssic, creds, ctx->flags, &req_body,
+		      input_token, output_token);
+
+    /*
+     * If FAST authenticated the KDC (which will be the case unless anonymous
```
Comment (nicowilliams) [inline lib/krb5/init_creds_pw.c:1251]: Only true if we're strengthening the reply key, not replacing it.

Comment (nicowilliams) [reply lib/krb5/init_creds_pw.c:1251]: Anyways, lets leave that for the Apple merge.

**Comment by nicowilliams** [comment]
LGTM!

**Comment by nicowilliams** [comment]
All code I review should be this good!

**Comment by lhoward** [comment]
> All code I review should be this good!

Hehe, you've made my day :)

**Comment by nicowilliams** [comment]
`gss_get_instance()` isn't needed because the GSS ABI, though it isn't standard, it might as well be, and we'll never break it.  I'm afraid my comment about there not being one led you astray.  You may remove it, or keep it -- I don't really care either way.

**Comment by lhoward** [comment]
> `gss_get_instance()` isn't needed because the GSS ABI, though it isn't standard, it might as well be, and we'll never break it. I'm afraid my comment about there not being one led you astray. You may remove it, or keep it -- I don't really care either way.

I suppose some non-standard extensions might change between releases or implementations…

**Pull request merged**